File: cpp/forge_msgs/src/imu.cpp
#include "detail.hpp"

#include <cmath>
#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

namespace forge_msgs {

using namespace detail;

namespace {

std::vector<std::shared_ptr<arrow::Field>> OrientationFields() {
  return {arrow::field("qx", arrow::float64(), false),
          arrow::field("qy", arrow::float64(), false),
          arrow::field("qz", arrow::float64(), false),
          arrow::field("qw", arrow::float64(), false)};
}

std::vector<std::shared_ptr<arrow::Field>> Vector3Fields() {
  return {arrow::field("x", arrow::float64(), false),
          arrow::field("y", arrow::float64(), false),
          arrow::field("z", arrow::float64(), false)};
}

arrow::Status ValidateFiniteVector(const std::string& name,
                                   const ImuVector3& value) {
  if (!std::isfinite(value.x) || !std::isfinite(value.y) ||
      !std::isfinite(value.z)) {
    return arrow::Status::Invalid(name, " components must be finite");
  }
  return arrow::Status::OK();
}

arrow::Status ValidateCovariance(const std::string& name,
                                 const std::vector<double>& values) {
  if (!values.empty() && values.size() != 9) {
    return arrow::Status::Invalid(name,
                                  " must be empty or contain exactly 9 values");
  }
  for (double value : values) {
    if (!std::isfinite(value)) {
      return arrow::Status::Invalid(name, " values must be finite");
    }
  }
  if (!values.empty() &&
      (values[0] < 0.0 || values[4] < 0.0 || values[8] < 0.0)) {
    return arrow::Status::Invalid(name,
                                  " diagonal values must be non-negative");
  }
  return arrow::Status::OK();
}

arrow::Result<std::shared_ptr<arrow::Array>> OrientationArray(
    const std::optional<ImuOrientation>& orientation) {
  const ImuOrientation stored = orientation.value_or(ImuOrientation{});
  ARROW_ASSIGN_OR_RAISE(auto qx, ScalarF64(stored.qx));
  ARROW_ASSIGN_OR_RAISE(auto qy, ScalarF64(stored.qy));
  ARROW_ASSIGN_OR_RAISE(auto qz, ScalarF64(stored.qz));
  ARROW_ASSIGN_OR_RAISE(auto qw, ScalarF64(stored.qw));
  ARROW_ASSIGN_OR_RAISE(
      auto batch,
      MakeBatch(OrientationFields(), {qx, qy, qz, qw}));
  return StructScalar(*batch, orientation.has_value());
}

arrow::Result<std::shared_ptr<arrow::Array>> Vector3Array(
    const ImuVector3& value) {
  ARROW_ASSIGN_OR_RAISE(auto x, ScalarF64(value.x));
  ARROW_ASSIGN_OR_RAISE(auto y, ScalarF64(value.y));
  ARROW_ASSIGN_OR_RAISE(auto z, ScalarF64(value.z));
  ARROW_ASSIGN_OR_RAISE(auto batch, MakeBatch(Vector3Fields(), {x, y, z}));
  return StructScalar(*batch);
}

arrow::Result<int> UniqueColumnIndex(const arrow::RecordBatch& batch,
                                     const std::string& name) {
  int index = -1;
  int count = 0;
  for (int candidate = 0; candidate < batch.num_columns(); ++candidate) {
    if (batch.schema()->field(candidate)->name() == name) {
      index = candidate;
      ++count;
    }
  }
  if (count == 0) {
    return arrow::Status::Invalid("missing ", name, " column");
  }
  if (count != 1) {
    return arrow::Status::Invalid(name,
                                  " column must appear exactly once, got ",
                                  count);
  }
  return index;
}

template <typename ArrayType>
arrow::Result<std::shared_ptr<ArrayType>> UniqueArray(
    const arrow::RecordBatch& batch, const std::string& name) {
  ARROW_ASSIGN_OR_RAISE(const int index, UniqueColumnIndex(batch, name));
  auto array = std::dynamic_pointer_cast<ArrayType>(batch.column(index));
  if (!array) {
    return arrow::Status::Invalid(name, " column has unexpected physical type");
  }
  if (array->length() != 1) {
    return arrow::Status::Invalid(name, " column must contain one row");
  }
  return array;
}

arrow::Status ValidateStructType(
    const std::shared_ptr<arrow::StructArray>& array,
    const std::vector<std::shared_ptr<arrow::Field>>& expected,
    const std::string& name) {
  const auto type = array->struct_type();
  if (type->num_fields() != static_cast<int>(expected.size())) {
    return arrow::Status::Invalid(name,
                                  " struct has an unexpected child count");
  }
  for (int index = 0; index < type->num_fields(); ++index) {
    const auto& expected_field = expected[static_cast<std::size_t>(index)];
    if (type->field(index)->name() != expected_field->name() ||
        !type->field(index)->type()->Equals(*expected_field->type())) {
      return arrow::Status::Invalid(
          name, " struct has unexpected child name, order, or physical type");
    }
  }
  return arrow::Status::OK();
}

arrow::Result<double> ReadStructDouble(
    const std::shared_ptr<arrow::StructArray>& array, int child_index,
    const std::string& name) {
  auto child =
      std::dynamic_pointer_cast<arrow::DoubleArray>(array->field(child_index));
  if (!child) {
    return arrow::Status::Invalid(name, " must be float64");
  }
  if (child->IsNull(0)) {
    return arrow::Status::Invalid(name, " must not be null");
  }
  return child->Value(0);
}

arrow::Result<std::optional<ImuOrientation>> ReadOrientation(
    const arrow::RecordBatch& batch) {
  ARROW_ASSIGN_OR_RAISE(
      auto array,
      UniqueArray<arrow::StructArray>(batch, "orientation"));
  ARROW_RETURN_NOT_OK(
      ValidateStructType(array, OrientationFields(), "orientation"));
  if (array->IsNull(0)) return std::optional<ImuOrientation>{};

  ImuOrientation value;
  ARROW_ASSIGN_OR_RAISE(value.qx,
                        ReadStructDouble(array, 0, "orientation.qx"));
  ARROW_ASSIGN_OR_RAISE(value.qy,
                        ReadStructDouble(array, 1, "orientation.qy"));
  ARROW_ASSIGN_OR_RAISE(value.qz,
                        ReadStructDouble(array, 2, "orientation.qz"));
  ARROW_ASSIGN_OR_RAISE(value.qw,
                        ReadStructDouble(array, 3, "orientation.qw"));
  return std::optional<ImuOrientation>{value};
}

arrow::Result<ImuVector3> ReadVector3(const arrow::RecordBatch& batch,
                                      const std::string& name) {
  ARROW_ASSIGN_OR_RAISE(auto array,
                        UniqueArray<arrow::StructArray>(batch, name));
  ARROW_RETURN_NOT_OK(ValidateStructType(array, Vector3Fields(), name));
  if (array->IsNull(0)) {
    return arrow::Status::Invalid(name, " must contain one non-null struct row");
  }

  ImuVector3 value;
  ARROW_ASSIGN_OR_RAISE(value.x,
                        ReadStructDouble(array, 0, name + ".x"));
  ARROW_ASSIGN_OR_RAISE(value.y,
                        ReadStructDouble(array, 1, name + ".y"));
  ARROW_ASSIGN_OR_RAISE(value.z,
                        ReadStructDouble(array, 2, name + ".z"));
  return value;
}

arrow::Result<std::vector<double>> ReadCovariance(
    const arrow::RecordBatch& batch, const std::string& name) {
  ARROW_ASSIGN_OR_RAISE(auto list,
                        UniqueArray<arrow::ListArray>(batch, name));
  if (list->IsNull(0)) {
    return arrow::Status::Invalid(name,
                                  " must contain one non-null list row");
  }
  auto values =
      std::dynamic_pointer_cast<arrow::DoubleArray>(list->value_slice(0));
  if (!values) {
    return arrow::Status::Invalid(name, " values must be float64");
  }
  if (values->null_count() != 0) {
    return arrow::Status::Invalid(name, " values must not contain nulls");
  }

  std::vector<double> result;
  result.reserve(static_cast<std::size_t>(values->length()));
  for (std::int64_t index = 0; index < values->length(); ++index) {
    result.push_back(values->Value(index));
  }
  return result;
}

arrow::Result<std::optional<double>> ReadTemperature(
    const arrow::RecordBatch& batch) {
  ARROW_ASSIGN_OR_RAISE(
      auto array,
      UniqueArray<arrow::DoubleArray>(batch, "temperature_celsius"));
  if (array->IsNull(0)) return std::optional<double>{};
  return std::optional<double>{array->Value(0)};
}

}  // namespace

arrow::Status ImuOrientation::Validate() const {
  if (!std::isfinite(qx) || !std::isfinite(qy) || !std::isfinite(qz) ||
      !std::isfinite(qw)) {
    return arrow::Status::Invalid(
        "orientation quaternion components must be finite");
  }
  if (qx == 0.0 && qy == 0.0 && qz == 0.0 && qw == 0.0) {
    return arrow::Status::Invalid(
        "orientation quaternion must not be all zero");
  }
  return arrow::Status::OK();
}

arrow::Status ImuVector3::Validate() const {
  return ValidateFiniteVector("ImuVector3", *this);
}

arrow::Status Imu::Validate() const {
  if (orientation) ARROW_RETURN_NOT_OK(orientation->Validate());
  ARROW_RETURN_NOT_OK(
      ValidateFiniteVector("angular_velocity", angular_velocity));
  ARROW_RETURN_NOT_OK(
      ValidateFiniteVector("linear_acceleration", linear_acceleration));
  ARROW_RETURN_NOT_OK(
      ValidateCovariance("orientation_covariance", orientation_covariance));
  ARROW_RETURN_NOT_OK(ValidateCovariance("angular_velocity_covariance",
                                        angular_velocity_covariance));
  ARROW_RETURN_NOT_OK(ValidateCovariance("linear_acceleration_covariance",
                                        linear_acceleration_covariance));
  if (!orientation && !orientation_covariance.empty()) {
    return arrow::Status::Invalid(
        "orientation_covariance must be empty when orientation is absent");
  }
  if (temperature_celsius && !std::isfinite(*temperature_celsius)) {
    return arrow::Status::Invalid(
        "temperature_celsius must be finite when present");
  }
  return arrow::Status::OK();
}

arrow::Result<std::shared_ptr<arrow::RecordBatch>> Imu::ToRecordBatch() const {
  ARROW_RETURN_NOT_OK(Validate());

  ARROW_ASSIGN_OR_RAISE(auto orientation_array, OrientationArray(orientation));
  ARROW_ASSIGN_OR_RAISE(auto angular_velocity_array,
                        Vector3Array(angular_velocity));
  ARROW_ASSIGN_OR_RAISE(auto linear_acceleration_array,
                        Vector3Array(linear_acceleration));
  ARROW_ASSIGN_OR_RAISE(auto orientation_covariance_array,
                        F64List(orientation_covariance));
  ARROW_ASSIGN_OR_RAISE(auto angular_velocity_covariance_array,
                        F64List(angular_velocity_covariance));
  ARROW_ASSIGN_OR_RAISE(auto linear_acceleration_covariance_array,
                        F64List(linear_acceleration_covariance));
  ARROW_ASSIGN_OR_RAISE(auto temperature_array,
                        OptionalF64(temperature_celsius));

  const auto orientation_type = arrow::struct_(OrientationFields());
  const auto vector_type = arrow::struct_(Vector3Fields());
  const auto covariance_type = ListType(arrow::float64());
  return MakeBatch(
      {arrow::field("orientation", orientation_type, true),
       arrow::field("angular_velocity", vector_type, false),
       arrow::field("linear_acceleration", vector_type, false),
       arrow::field("orientation_covariance", covariance_type, false),
       arrow::field("angular_velocity_covariance", covariance_type, false),
       arrow::field("linear_acceleration_covariance", covariance_type, false),
       arrow::field("temperature_celsius", arrow::float64(), true)},
      {orientation_array, angular_velocity_array, linear_acceleration_array,
       orientation_covariance_array, angular_velocity_covariance_array,
       linear_acceleration_covariance_array, temperature_array});
}

arrow::Result<Imu> Imu::FromRecordBatch(const arrow::RecordBatch& batch) {
  ARROW_RETURN_NOT_OK(RequireOneRow(batch));
  Imu value;
  ARROW_ASSIGN_OR_RAISE(value.orientation, ReadOrientation(batch));
  ARROW_ASSIGN_OR_RAISE(value.angular_velocity,
                        ReadVector3(batch, "angular_velocity"));
  ARROW_ASSIGN_OR_RAISE(value.linear_acceleration,
                        ReadVector3(batch, "linear_acceleration"));
  ARROW_ASSIGN_OR_RAISE(value.orientation_covariance,
                        ReadCovariance(batch, "orientation_covariance"));
  ARROW_ASSIGN_OR_RAISE(
      value.angular_velocity_covariance,
      ReadCovariance(batch, "angular_velocity_covariance"));
  ARROW_ASSIGN_OR_RAISE(
      value.linear_acceleration_covariance,
      ReadCovariance(batch, "linear_acceleration_covariance"));
  ARROW_ASSIGN_OR_RAISE(value.temperature_celsius, ReadTemperature(batch));
  ARROW_RETURN_NOT_OK(value.Validate());
  return value;
}

}  // namespace forge_msgs

File: packages/msgs/src/forge_msgs/imu.py
from __future__ import annotations

import math
from typing import ClassVar, Self

import pyarrow as pa
from pydantic import BaseModel, Field, model_validator

from forge_msgs._arrow_read import (
    ArrowInput,
    normalize_single_row,
    required_field_indices,
)

_ORIENTATION_TYPE = pa.struct(
    [
        pa.field("qx", pa.float64(), nullable=False),
        pa.field("qy", pa.float64(), nullable=False),
        pa.field("qz", pa.float64(), nullable=False),
        pa.field("qw", pa.float64(), nullable=False),
    ]
)
_VECTOR3_TYPE = pa.struct(
    [
        pa.field("x", pa.float64(), nullable=False),
        pa.field("y", pa.float64(), nullable=False),
        pa.field("z", pa.float64(), nullable=False),
    ]
)
_COVARIANCE_TYPE = pa.list_(pa.float64())
_IMU_ARROW_SCHEMA = pa.schema(
    [
        pa.field("orientation", _ORIENTATION_TYPE, nullable=True),
        pa.field("angular_velocity", _VECTOR3_TYPE, nullable=False),
        pa.field("linear_acceleration", _VECTOR3_TYPE, nullable=False),
        pa.field("orientation_covariance", _COVARIANCE_TYPE, nullable=False),
        pa.field("angular_velocity_covariance", _COVARIANCE_TYPE, nullable=False),
        pa.field("linear_acceleration_covariance", _COVARIANCE_TYPE, nullable=False),
        pa.field("temperature_celsius", pa.float64(), nullable=True),
    ]
)


class ImuOrientation(BaseModel):
    """Quaternion orientation with explicitly named XYZW components."""

    qx: float
    qy: float
    qz: float
    qw: float

    @model_validator(mode="after")
    def _validate(self) -> Self:
        values = (self.qx, self.qy, self.qz, self.qw)
        if any(not math.isfinite(value) for value in values):
            raise ValueError("Imu orientation components must be finite")
        if all(value == 0.0 for value in values):
            raise ValueError("Imu orientation quaternion must not be all zero")
        return self


class ImuVector3(BaseModel):
    """Three-axis SI-unit inertial vector."""

    x: float
    y: float
    z: float

    @model_validator(mode="after")
    def _validate(self) -> Self:
        if any(not math.isfinite(value) for value in (self.x, self.y, self.z)):
            raise ValueError("Imu vector components must be finite")
        return self


def _validate_covariance(name: str, values: list[float]) -> None:
    if len(values) not in (0, 9):
        raise ValueError(f"Imu {name} must be empty or contain exactly 9 values")
    if any(not math.isfinite(value) for value in values):
        raise ValueError(f"Imu {name} values must be finite")
    if values and any(values[index] < 0.0 for index in (0, 4, 8)):
        raise ValueError(f"Imu {name} diagonal values must be non-negative")


def _read_struct(
    batch: pa.RecordBatch,
    index: int,
    name: str,
    child_names: tuple[str, ...],
    *,
    nullable: bool,
) -> dict[str, float] | None:
    cell = batch.column(index)[0]
    if not cell.is_valid:
        if nullable:
            return None
        raise ValueError(f"Imu Arrow field {name} struct cell must not be null")

    values: dict[str, float] = {}
    for child_name in child_names:
        child = cell[child_name]
        if not child.is_valid:
            raise ValueError(
                f"Imu Arrow field {name} struct child {child_name} must not be null"
            )
        values[child_name] = float(child.as_py())
    return values


def _read_covariance(batch: pa.RecordBatch, index: int, name: str) -> list[float]:
    cell = batch.column(index)[0]
    if not cell.is_valid:
        raise ValueError(f"Imu Arrow field {name} list cell must not be null")
    values = cell.values
    if values.null_count:
        raise ValueError(f"Imu Arrow field {name} list items must not be null")
    return [float(value) for value in values.to_pylist()]


def _read_optional_float(batch: pa.RecordBatch, index: int) -> float | None:
    cell = batch.column(index)[0]
    if not cell.is_valid:
        return None
    return float(cell.as_py())


class Imu(BaseModel):
    """Single SI-unit inertial measurement sample."""

    _ARROW_SCHEMA: ClassVar[pa.Schema] = _IMU_ARROW_SCHEMA

    orientation: ImuOrientation | None = None
    angular_velocity: ImuVector3
    linear_acceleration: ImuVector3
    orientation_covariance: list[float] = Field(default_factory=list)
    angular_velocity_covariance: list[float] = Field(default_factory=list)
    linear_acceleration_covariance: list[float] = Field(default_factory=list)
    temperature_celsius: float | None = None

    @model_validator(mode="after")
    def _validate(self) -> Self:
        _validate_covariance("orientation_covariance", self.orientation_covariance)
        _validate_covariance(
            "angular_velocity_covariance", self.angular_velocity_covariance
        )
        _validate_covariance(
            "linear_acceleration_covariance", self.linear_acceleration_covariance
        )
        if self.orientation is None and self.orientation_covariance:
            raise ValueError(
                "Imu orientation_covariance must be empty when orientation is null"
            )
        if self.temperature_celsius is not None and not math.isfinite(
            self.temperature_celsius
        ):
            raise ValueError("Imu temperature_celsius must be finite when present")
        return self

    def to_arrow(self) -> pa.RecordBatch:
        validated = type(self).model_validate(self.model_dump(mode="python"))
        orientation = (
            None
            if validated.orientation is None
            else validated.orientation.model_dump(mode="python")
        )
        arrays: list[pa.Array] = [
            pa.array([orientation], type=_ORIENTATION_TYPE),
            pa.array(
                [validated.angular_velocity.model_dump(mode="python")],
                type=_VECTOR3_TYPE,
            ),
            pa.array(
                [validated.linear_acceleration.model_dump(mode="python")],
                type=_VECTOR3_TYPE,
            ),
            pa.array([validated.orientation_covariance], type=_COVARIANCE_TYPE),
            pa.array([validated.angular_velocity_covariance], type=_COVARIANCE_TYPE),
            pa.array([validated.linear_acceleration_covariance], type=_COVARIANCE_TYPE),
            pa.array([validated.temperature_celsius], type=pa.float64()),
        ]
        return pa.RecordBatch.from_arrays(arrays, schema=self._ARROW_SCHEMA)

    @classmethod
    def from_arrow(cls, data: ArrowInput) -> Imu:
        batch = normalize_single_row(data, "Imu")
        indices = required_field_indices(batch, _IMU_ARROW_SCHEMA, "Imu")
        orientation_values = _read_struct(
            batch,
            indices["orientation"],
            "orientation",
            ("qx", "qy", "qz", "qw"),
            nullable=True,
        )
        angular_velocity_values = _read_struct(
            batch,
            indices["angular_velocity"],
            "angular_velocity",
            ("x", "y", "z"),
            nullable=False,
        )
        linear_acceleration_values = _read_struct(
            batch,
            indices["linear_acceleration"],
            "linear_acceleration",
            ("x", "y", "z"),
            nullable=False,
        )
        assert angular_velocity_values is not None
        assert linear_acceleration_values is not None
        return cls(
            orientation=(
                None
                if orientation_values is None
                else ImuOrientation.model_validate(orientation_values)
            ),
            angular_velocity=ImuVector3.model_validate(angular_velocity_values),
            linear_acceleration=ImuVector3.model_validate(linear_acceleration_values),
            orientation_covariance=_read_covariance(
                batch,
                indices["orientation_covariance"],
                "orientation_covariance",
            ),
            angular_velocity_covariance=_read_covariance(
                batch,
                indices["angular_velocity_covariance"],
                "angular_velocity_covariance",
            ),
            linear_acceleration_covariance=_read_covariance(
                batch,
                indices["linear_acceleration_covariance"],
                "linear_acceleration_covariance",
            ),
            temperature_celsius=_read_optional_float(
                batch, indices["temperature_celsius"]
            ),
        )
